### ISSUE TITLE
question about multiple watchers

### ISSUE THREAD

**Issue by szymonkaliski** [Hammerspoon/hammerspoon#1245]
My HS config is by now quite big, and I'm using the same watchers in multiple places, for example battery watcher in one file is displaying notifications about percentage, in another file I have another battery watcher instance which is automatically unmount usb drives, etc... 

My question is could that possibly impact my old macbook performance, and would it be better to just have one watcher per each watcher type that I'm using, or is having multiple instances fine?

**Comment by asmagill** [comment]
The simple answer is yes, they do add up.  Which ones you should worry about most... that's a somewhat tougher question.

tl;dr - this is on my mind as well and I am working out some ideas that should hopefully help reduce the impact of this in the future.  At a minimum, I recommend taking a look at `hs.watchable` and see if that can help you with setting up single watchers that can trigger multiple callbacks.

### My thoughts and plans, FWIW

Consider what is happening when a watcher/notification/eventtap is triggered (this is very simplified and generic and I'm being loose with some of the terminology, so it may not be exact for any particular given watcher):
* first there is the "event" which causes the macOS to generate the "notification"
* Hammerspoon is notified/detects the change which causes a context shift to the Hammerspoon application (i.e. notable overhead)
* If the logic in the compiled portion of Hammerspoon believes that the callback should occur, the Lua state is invoked (more notable overhead)
* The callback returns, Hammerspoon shifts context back to the OS
* If multiple watchers for the same thing are registered, go back to step 2 for the next one

Some random thoughts of my own regarding this and some of the things I am working on related to it:

* This was part of the rationale behind `hs.watchable` which allows you to create a path that other code can subscribe to.  When the watcher/code which "owns" the path changes its value, it triggers updates to the subscribers through Lua metamethods.  Yes, we stay in Lua longer, which is arguably the slowest stage, but it means we only need one watcher (and thus context shift) for however many essentially identical watchers we want

* Push as much of the logic to the operating system or compiled code as possible:  For example, when using `hs.distributednotifications.new`, it is significantly more efficient to specify a `name` (and when possible `object`) parameter rather then use a series of if-then comparisons in the callback to determine if we actually care about the notification.

* `hs.wifi.watcher` was recently rewritten because the way macOS notifies Hammerspoon of changes was causing callbacks to be fired multiple times for the same event.  The new model only registers one watcher per event and when you create a new watcher, a reference to the new callback is added to an array Hammerspoon maintains internally -- i.e. Hammerspoon only gets notified once now for an SSID change, but Hammerspoon then knows to perform N number of callbacks for that one event.  It is this model that I am migrating `hs.eventtap` to though it probably won't be available for testing until later this month.

* While this is still in the early investigation stages, I am also thinking about ways we can push some of the logic for determining if Hammerspoon even cares about an event into the compiled portion of Hammerspoon... the simplest example I can think of is an `hs.eventtap` watcher which is looking for `keyDown` events for `h`, `j`, `k`, and `l`... since we know we don't care about any other letter, it would be faster if the Objective-C portion of Hammerspoon could determine this rather then rely on the Lua callback to check the keycodes for us... the trick is making this "specification language" easy to parse both by the computer and by us humans that have to write it and yet be flexible enough to cover enough of the conditions we care about and become the rule rather than the exception (e.g. we want the callback for the above letters if the `fn` modifier is down but otherwise, like other letters, we don't need to invoke the callback... but we'd want a syntax that could easily specify that we want *only* the `fn` modifier while allowing another "watcher" to specify that we want the callback as long as the *combination* of modifiers includes the `fn` modifier)

So you are correct in that they add up, but over the next couple of months I hope to add some modules and changes that will help mitigate this and reduce the number of independent (but essentially identical as far as the operating system is concerned) notifications coming into Hammerspoon.

**Comment by szymonkaliski** [comment]
@asmagill thank you for this amazing answer, I will rewrite my watchers to use hs.watchable for now, and I'm looking forward to the upcoming updates!

**Comment by cmsj** [comment]
(agreed that @asmagill 's answer here is awesome. I'm gonna close this issue out because I think the question has been answered, and there isn't any direct work to be done that isn't already being done :)

re-open if you disagree.
